File: Medium/200 - NumberOfIslands/Solution.cs
/// Question 200. Number of Islands (https://leetcode-cn.com/problems/number-of-islands/)
/// 执行用时 : 144 ms, 在所有 C# 提交中击败了 26.06% 的用户
/// 内存消耗 : 27.7 MB, 在所有 C# 提交中击败了 100.00% 的用户

public class Solution
{
    public int NumIslands(char[][] grid)
    {
        List<(int, int)> directions = new List<(int, int)>() { (-1, 0), (1, 0), (0, 1), (0, -1) };

        int gridRows = grid.Count();
        if (gridRows == 0) { return 0; }

        int gridColumns = grid.First().Count();
        List<(int, int)> visitedLands = new List<(int, int)>();
        int islandCount = 0;

        for (int i = 0; i < gridRows; i++)
        {
            for (int j = 0; j < gridColumns; j++)
            {
                if (grid[i][j].Equals('1'))
                {
                    islandCount++;

                    Queue<(int, int)> adjacentLands = new Queue<(int, int)>();
                    adjacentLands.Enqueue((i, j));

                    while (adjacentLands.Count() > 0)
                    {
                        (int landX, int landY) = adjacentLands.Dequeue();

                        for (int k = 0; k < directions.Count(); k++)
                        {
                            (int directionX, int directionY) = directions.ElementAt(k);

                            int newX = landX + directionX;
                            int newY = landY + directionY;

                            if (newX < 0 || newY < 0 ||
                                newX >= gridRows || newY >= gridColumns ||
                                grid[newX][newY].Equals('0'))
                            {
                                continue;
                            }

                            grid[newX][newY] = '0';
                            adjacentLands.Enqueue((newX, newY));
                        }
                    }
                }
            }
        }


        return islandCount;
    }
}

File: Medium/200 - NumberOfIslands/solution.ts
/// Question 200. Number of Islands (https://leetcode-cn.com/problems/number-of-islands/)

const DIRECTIONS: number[][] = [
  [-1, 0],
  [1, 0],
  [0, 1],
  [0, -1],
];

/**
 * 执行用时：96 ms, 在所有 typescript 提交中击败了 46.91% 的用户
 * 内存消耗： 40.8 MB, 在所有 typescript 提交中击败了 17.86% 的用户
 *
 * @param {string[][]} grid
 * @returns {number}
 */
function numIslands(grid: string[][]): number {
  if (grid.length === 0) {
    return 0;
  }

  const ROW_COUNT: number = grid.length;
  const COLUMN_COUNT: number = grid[0].length;

  let islandCount: number = 0;
  let adjacentLands: number[][] = [];

  for (let i = 0; i < ROW_COUNT; i++) {
    for (let j = 0; j < COLUMN_COUNT; j++) {
      if (grid[i][j] === "1") {
        islandCount++;
        adjacentLands = [[i, j]];

        while (adjacentLands.length) {
          const [adjacentX, adjacentY] = adjacentLands.shift() as number[];

          for (const [xDirection, yDirection] of DIRECTIONS) {
            const newX: number = adjacentX + xDirection;
            const newY: number = adjacentY + yDirection;

            if (
              newX < 0 ||
              newX >= ROW_COUNT ||
              newY < 0 ||
              newY >= COLUMN_COUNT ||
              grid[newX][newY] === "0"
            ) {
              continue;
            }

            adjacentLands.push([newX, newY]);
            grid[newX][newY] = "0";
          }
        }
      }
    }
  }

  return islandCount;
}
